client_linux.go

```diff
@@ -139,3 +163,60 @@ func parsePorts(msgs []genetlink.Message) ([]*Port, error) {
 
 	return ps, nil
 }
+
+// parseDpipeTables parses DPIPE tables from a slice of generic netlink messages.
+func parseDpipeTables(msgs []genetlink.Message) ([]*DpipeTable, error) {
+	var bus, dev string
+	if len(msgs) == 0 {
+		// No devlink response found.
+		return nil, nil
+	}
+
+	ts := make([]*DpipeTable, 0, len(msgs))
+	for _, m := range msgs {
+		ad, err := netlink.NewAttributeDecoder(m.Data)
+		if err != nil {
+			return nil, err
+		}
+
+		for ad.Next() {
+			switch ad.Type() {
+			case unix.DEVLINK_ATTR_BUS_NAME:
+				bus = ad.String()
+			case unix.DEVLINK_ATTR_DEV_NAME:
+				dev = ad.String()
+			case unix.DEVLINK_ATTR_DPIPE_TABLES:
+				tablesData := ad.Bytes()
+				adTables, err := netlink.NewAttributeDecoder(tablesData)
+				if err != nil {
+					continue
+				}
+				for adTables.Next() {
+					if adTables.Type() == unix.DEVLINK_ATTR_DPIPE_TABLE {
+						tableData := adTables.Bytes()
+						adTable, err := netlink.NewAttributeDecoder(tableData)
+						if err != nil {
+							continue
+						}
+						var t DpipeTable
+						t.Bus = bus
+						t.Device = dev
+						for adTable.Next() {
+							switch adTable.Type() {
+							case unix.DEVLINK_ATTR_DPIPE_TABLE_NAME:
+								t.Name = adTable.String()
+							case unix.DEVLINK_ATTR_DPIPE_TABLE_SIZE:
+								t.Size = adTable.Uint64()
+							case unix.DEVLINK_ATTR_DPIPE_TABLE_COUNTERS_ENABLED:
+								t.CountersEnabled = adTable.Uint8() != 0
+							}
+						}
+						ts = append(ts, &t)
+					}
+				}
+			}
+
+		}
```

Comment:
Need to check `ad.Err()` at the end of the `ad.Next()` loop or errors will not be detected.

Reply:
Fixed.